### ISSUE TITLE
AuthData structure (aka Blob v2)

### ISSUE THREAD

**Issue by Akuukis** [stellarchat/desktop-client#255]
The latest PR (#253) introduces various features, as diverse signing methods and awareness of multi-signing. The current Blob structure doesn't support neither of them, thus I propose Blob v2 called AuthData. Here's the main points:
- due multisigning, the account may have multiple signers. Thus, instead of single "masterkey" string it now has "keypairs" array.
- due diverse signing methods, the secret may be stored externally, i.e. in hardwallet. Nevetheless it's important to save publicKey of such keypair to be able to recognize it as owned, thus "Keypair" is a object, instead of plain secret.
- due both of them, signing method is defined per Keypair instead of generally, because one may want to store his secrets with different means.
- added versioning
- internal structure of AuthData instance will be based on this interface.

I am looking forward to implement this in next days.

```typescript
// Version 2

// Keystore version 3 as defined here: https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition
// ... and implemented here: https://github.com/ethereumjs/ethereumjs-wallet/blob/master/index.js
interface Keystore {
  version: number,  // uses 3
  id: string,
  crypto: {
    ciphertext: string,
    cipher: string,  // uses 'aes-128-ctr'
    cipherparams: CipherParams,  // uses { iv: string },
    kdf: string,  // uses 'scrypt'
    kdfparams: KdfParams,  // uses  { dklen: number, salt: string, n: number, r: number, p: number },
    mac: string,
  }
  // Non-standard extension ahead.
  content: 'stellar',  // Signal that content is Stellar technology's AuthData, not a plain Ethereum private key.
  contentparams: {
      version: '2.0',  // Decimal point mandatory. First digit is major (breaking) change, second is minor (compatible)
  }
}

interface AuthData {
    address : string  // Verified with `StellarSdk.isValidEd25519PublicKey(address)`
    network : string  // Network passphrase of network where the account is created in
    keypairs: Keypair[]  // To know what publicKeys user owns lets UI to decide how to request their signatures.
    contacts: Contact[]  // Sorted by time created.
    created : string  // local timezone, ISO format: 2018-06-13T08:13:56+00:00
    updated : string  // local timezone, ISO format: 2018-06-13T08:13:56+00:00
}

interface Keypair {
    publicKey    : string  // Verified with `StellarSdk.isValidEd25519PublicKey(address)`
    signingMethod: SigningMethod  // Force user to sign in one way only.
    details      : string  // depends on signingMethod, i.e. for "encryptedSecret" that's the secret itself.
}

enum SigningMethod {
    ENCRYPTED_SECRET = 'encryptedSecret',
    EXTERNAL_SECRET  = 'externalSecret',
    COLD_MACHINE     = 'coldMachine',
    HARDWARE_WALLET  = 'hardwareWallet',
    UNKNOWN          = 'unknown',  // Effectively everything at the same time. Default value until used first time.
    NONE             = 'none',  // Effectively read-only. Remind user that he can't sign that.
}

interface Contact {
    address    : string | undefined  // at least one of "address" or "federation" must be defined.
    federation : string | undefined  // If both of them are, "address" takes precedence.
    name       : string  // UTF-8, max 100 characters, must be at least 1 character.
    details    : string  // UTF-8, unlimited characters?, may be empty.
    created    : string  // local timezone, ISO format: 2018-06-13T08:13:56+00:00
    updated    : string  // local timezone, ISO format: 2018-06-13T08:13:56+00:00
    // Memo is verified with `try{ new StellarSdk.Memo(type, value); return true} catch(e) { return false}`
    defaultMemoType : 'none' | 'id' | 'text' | 'hash' | 'return'  // Matches constants as in `StellarSdk.Memo.*`
    defaultMemoValue: undefined | string | number  // According to memo type
}

function theGistOfEncryption(authData, password) {
    const cipherparams = { iv: require('crypto').randomBytes(16) };
    const kdfparams = { dklen: 32, salt: require('crypto').randomBytes(32), n: 262144, r: 8, p: 1 };
    const plaintext = JSON.stringify(authData);
    const id = require('uuid/v4')({ random: require('crypto').randomBytes(16) });
    const derivedKey = require('scryptsy')(Buffer.from(password), kdfparams.salt, kdfparams.n, kdfparams.r, kdfparams.p, kdfparams.dklen);
    const cipher = require('crypto').createCipheriv('aes-128-ctr', derivedKey.slice(0, 16), cipherparams.iv);
    const ciphertext = Buffer.concat([ cipher.update(plaintext), cipher.final() ]);
    const mac = require('keccak256')(Buffer.concat([ derivedKey.slice(16, 32), Buffer.from(ciphertext, 'hex') ]));
    const fileContents = JSON.stringify({
        version: 3,
        id: id,
        content: 'stellar',
        contentparams: {
            version: '2.0',
        },
        crypto: {
            ciphertext: ciphertext.toString('hex'),
            cipher: 'aes-128-ctr',
            cipherparams: cipherparams,
            kdf: 'scrypt',
            kdfparams: kdfparams,
            mac: mac.toString('hex'),
        },
    });
    return fileContents;
}

// Using syntax of Typescript, because that's closest to JS and has interfaces and enums.
```

Here's a rough picture of current structure for comparison. Perhaps @stellarchat can correct me on this.

```typescript
// Version 1

interface Contact {
    name      : string
    view      : string
    address   : string
    memotype? : 'id' | 'text' | 'hash'
    memo?     : string
}

interface Blob {
    account_id: string
    masterkey: string
    contacts: Contact[]
    created: string  // (new Date()).toString() format.
}

function theGistOfEncryption(blob, password) {
    const plaintext = JSON.stringify(blob);
    const ciphertext = require('sjcl').encrypt(`${password.length}|${password}`, plaintext, {ks: 256, iter: 1000});
    const fileContents = btoa(ciphertext);
    return fileContents;
}
```

P.S. The syntax used is Typescript, because that's closest to JS and has interfaces and enums.

P.P.S. Heads up to biggest contributors @stellarchat @ripplefox @liangran

**Comment by stellarchat** [comment]
@Akuukis The current blob (version 1) structure is correct. It is not suitable for multi-signed account. Thus version 2 is necessary if more feature needed. 

The key point is backward compatibility. The wallet should make sure user would not lost their keys when upgrade the wallet.

**Comment by Akuukis** [comment]
@stellarchat Thank you for confirmation. I agree that backward compatibility is a must and it will be implemented too. Although I think it's safe to "auto-upgrade" all blobs to v2, while keeping a copy of v1 blob.

**Comment by Akuukis** [comment]
Update: `AuthData` should also have property `network` (with either network passphrase value) to signal on which network the account is created on.

**Comment by Akuukis** [comment]
Update: So far we have discussed only data structure, but we can also do better at encrypting that data. I will adapt Web3 Keystore structure version 3 for that. Edited first message.

**Comment by Akuukis** [comment]
This is implemented on top of #265 branch, so as soon it's merged I will open a new PR for this issue's implementation.
